quic/s2n-quic-transport/src/stream/controller.rs

```diff
@@ -69,122 +62,169 @@ impl Controller {
-                ),
-                remote_initiated: RemoteInitiated::new(initial_local_limits.max_streams_uni),
-            },
+            local_bidi_controller: LocalInitiated::new(
+                initial_peer_limits.max_open_remote_bidirectional_streams,
+                stream_limits.max_open_local_bidirectional_streams,
+            ),
+            remote_bidi_controller: RemoteInitiated::new(
+                initial_local_limits.max_open_remote_bidirectional_streams,
+            ),
+            local_uni_controller: LocalInitiated::new(
+                initial_peer_limits.max_open_remote_unidirectional_streams,
+                stream_limits.max_open_local_unidirectional_streams,
+            ),
+            remote_uni_controller: RemoteInitiated::new(
+                initial_local_limits.max_open_remote_unidirectional_streams,
+            ),
         }
     }
 
     /// This method is called when a `MAX_STREAMS` frame is received,
     /// which signals an increase in the available streams budget.
     pub fn on_max_streams(&mut self, frame: &MaxStreams) {
         match frame.stream_type {
-            StreamType::Bidirectional => self.bidi_controller.local_initiated.on_max_streams(frame),
-            StreamType::Unidirectional => self.uni_controller.local_initiated.on_max_streams(frame),
+            StreamType::Bidirectional => self.local_bidi_controller.on_max_streams(frame),
+            StreamType::Unidirectional => self.local_uni_controller.on_max_streams(frame),
         }
     }
 
-    /// This method is called when the local application wishes to open a new stream.
+    /// This method is called when the local application wishes to open the next stream
+    /// of a type (Bidirectional/Unidirectional).
     ///
     /// `Poll::Pending` is returned when there isn't available capacity to open a stream,
     /// either because of local initiated concurrency limits or the peer's stream limits.
     /// If `Poll::Pending` is returned, the waker in the given `context` will be woken
     /// when additional stream capacity becomes available.
-    pub fn poll_local_open_stream(
+    pub fn poll_open_local_stream(
         &mut self,
         stream_type: StreamType,
         open_tokens: &mut connection::OpenToken,
         context: &Context,
     ) -> Poll<()> {
-        match stream_type {
+        let poll_open = match stream_type {
             StreamType::Bidirectional => self
-                .bidi_controller
-                .local_initiated
+                .local_bidi_controller
                 .poll_open_stream(&mut open_tokens.bidirectional, context),
             StreamType::Unidirectional => self
-                .uni_controller
-                .local_initiated
+                .local_uni_controller
                 .poll_open_stream(&mut open_tokens.unidirectional, context),
-        }
+        };
+
+        // returns Pending if there is no capacity available
+        ready!(poll_open);
```

Comment:
Think you can probably get rid of `poll_open` and wrap the match in `ready!`, not a big deal though